File: push/src/test/java/io/appmetrica/analytics/push/impl/utils/AppMetricaTrackerTest.java
package io.appmetrica.analytics.push.impl.utils;

import android.content.Context;
import io.appmetrica.analytics.AppMetrica;
import io.appmetrica.analytics.IReporter;
import io.appmetrica.analytics.push.BuildConfig;
import io.appmetrica.analytics.push.coreutils.internal.CoreConstants;
import io.appmetrica.analytics.push.impl.AppMetricaPushCore;
import io.appmetrica.analytics.push.impl.PushServiceControllerComposite;
import io.appmetrica.analytics.push.testutils.MockedStaticRule;
import io.appmetrica.analytics.push.testutils.RandomStringGenerator;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class AppMetricaTrackerTest {

    @Rule
    public final MockedStaticRule<AppMetrica> appMetricaMockedStaticRule = new MockedStaticRule<>(AppMetrica.class);
    @Rule
    public final MockedStaticRule<AppMetricaPushCore> sAppMetricaPushCore =
        new MockedStaticRule<>(AppMetricaPushCore.class);

    private Context mContext;
    private IReporter mReporter;
    private AppMetricaTracker mTracker;
    private AppMetricaPushCore core;
    private PushServiceControllerComposite controller;

    @Before
    public void setUp() {
        mContext = RuntimeEnvironment.application;
        mReporter = mock(IReporter.class);

        core = mock(AppMetricaPushCore.class);
        controller = mock(PushServiceControllerComposite.class);
        when(AppMetrica.getReporter(any(Context.class), anyString())).thenReturn(mReporter);
        when(AppMetricaPushCore.getInstance(any(Context.class))).thenReturn(core);
        doReturn(controller).when(core).getPushServiceController();
        doReturn(Arrays.asList(CoreConstants.Transport.FIREBASE, CoreConstants.Transport.UNKNOWN))
            .when(controller).getTransportIds();
        mTracker = new AppMetricaTracker(mContext, new RandomStringGenerator().nextString());
    }

    @Test
    public void testResumeSessionShouldSendResumeToReporter() {
        mTracker.resumeSession();
        verify(mReporter, times(1)).resumeSession();
    }

    @Test
    public void testPauseSessionShouldSendResumeToReporter() {
        mTracker.pauseSession();
        verify(mReporter, times(1)).pauseSession();
    }

    @Test
    public void testReportEventShouldSendEventNameToReporter() {
        String eventName = new RandomStringGenerator().nextString();
        mTracker.reportEvent(eventName);
        ArgumentCaptor<String> arg = ArgumentCaptor.forClass(String.class);
        verify(mReporter, times(1)).reportEvent(arg.capture(), ArgumentMatchers.<String, Object>anyMap());
        assertThat(arg.getValue()).isEqualTo(eventName);
    }

    @Test
    public void testReportEventShouldIncludeSdkVersionToEnvironment() {
        mTracker.reportEvent(new RandomStringGenerator().nextString());
        ArgumentCaptor<Map> arg = ArgumentCaptor.forClass(Map.class);
        verify(mReporter, times(1)).reportEvent(anyString(), arg.capture());
        verifyConstainsVersion(arg.getValue());
    }

    @Test
    public void testReportEventWithAttributesShouldIncludeSdkVersionToEnvironment() {
        mTracker.reportEvent(new RandomStringGenerator().nextString(), new HashMap<String, Object>());
        ArgumentCaptor<Map> arg = ArgumentCaptor.forClass(Map.class);
        verify(mReporter, times(1)).reportEvent(anyString(), arg.capture());
        verifyConstainsVersion(arg.getValue());
    }

    private void verifyConstainsVersion(Map attributes) {
        assertThat(attributes).containsKey(AppMetricaTracker.SDK_VERSION_CODE_FIELD);
        assertThat(attributes.get(AppMetricaTracker.SDK_VERSION_CODE_FIELD))
            .isEqualTo(String.valueOf(BuildConfig.VERSION_CODE));
        assertThat(attributes).containsKey(AppMetricaTracker.TRANSPORT_FIELD);
        assertThat(attributes.get(AppMetricaTracker.TRANSPORT_FIELD))
            .isEqualTo("[" + CoreConstants.Transport.FIREBASE + ", " + CoreConstants.Transport.UNKNOWN + "]");
    }

    @Test
    public void testReportEventWithAttributesShouldSendEventNameToReporter() {
        String eventName = new RandomStringGenerator().nextString();
        mTracker.reportEvent(eventName, null);
        ArgumentCaptor<String> arg = ArgumentCaptor.forClass(String.class);
        verify(mReporter, times(1)).reportEvent(arg.capture(), ArgumentMatchers.<String, Object>anyMap());
        assertThat(arg.getValue()).isEqualTo(eventName);
    }

    @Test
    public void testReportEventWithAttributesShouldSendAttributesToReporter() {
        Map<String, Object> attributes = new HashMap<String, Object>();
        String key = new RandomStringGenerator().nextString();
        String value = new RandomStringGenerator().nextString();
        attributes.put(key, value);
        mTracker.reportEvent(new RandomStringGenerator().nextString(), attributes);
        ArgumentCaptor<Map> arg = ArgumentCaptor.forClass(Map.class);
        verify(mReporter, times(1)).reportEvent(anyString(), arg.capture());

        assertThat(arg.getValue()).containsKey(key);
        assertThat(arg.getValue().get(key)).isEqualTo(value);
    }

    @Test
    public void testReportErrorShouldSendMessageToReporter() {
        String message = new RandomStringGenerator().nextString();
        mTracker.reportError(message, null);
        ArgumentCaptor<String> arg = ArgumentCaptor.forClass(String.class);
        verify(mReporter, times(1)).reportError(arg.capture(), nullable(Throwable.class));
        assertThat(arg.getValue()).contains(message);
    }

    @Test
    public void testReportErrorShouldSendThrowableToReporter() {
        Throwable throwable = mock(Throwable.class);
        mTracker.reportError(null, throwable);
        ArgumentCaptor<Throwable> arg = ArgumentCaptor.forClass(Throwable.class);
        verify(mReporter, times(1)).reportError(anyString(), arg.capture());
        assertThat(arg.getValue()).isEqualTo(throwable);
    }

    @Test
    public void testReportErrorShouldContainsSdkVersionInErrorMessage() {
        String message = new RandomStringGenerator().nextString();
        mTracker.reportError(message, null);
        ArgumentCaptor<String> arg = ArgumentCaptor.forClass(String.class);
        verify(mReporter, times(1)).reportError(arg.capture(), nullable(Throwable.class));
        assertThat(arg.getValue())
            .contains(String.format("%s = %s", AppMetricaTracker.SDK_VERSION_CODE_FIELD, BuildConfig.VERSION_CODE));
//        assertThat(arg.getValue())
//            .contains(String.format("%s = %s", AppMetricaTracker.TRANSPORT_FIELD, AppMetricaPushCore.getProviderTitle()));
    }

    @Test
    public void testReportUnhandledExceptionShouldSendThrowableToReporter() {
        Throwable throwable = mock(Throwable.class);
        mTracker.reportUnhandledException(throwable);
        ArgumentCaptor<Throwable> arg = ArgumentCaptor.forClass(Throwable.class);
        verify(mReporter, times(1)).reportUnhandledException(arg.capture());
        assertThat(arg.getValue()).isEqualTo(throwable);
    }
}

File: push/src/test/java/io/appmetrica/analytics/push/impl/utils/AppMetricaTrackerTest.kt
package io.appmetrica.analytics.push.impl.utils

import android.content.Context
import io.appmetrica.analytics.AppMetrica
import io.appmetrica.analytics.IReporter
import io.appmetrica.analytics.push.BuildConfig
import io.appmetrica.analytics.push.coreutils.internal.CoreConstants
import io.appmetrica.analytics.push.impl.AppMetricaPushCore
import io.appmetrica.analytics.push.impl.PreferenceManager
import io.appmetrica.analytics.push.impl.PushServiceControllerComposite
import io.appmetrica.gradle.testutils.CommonTest
import io.appmetrica.gradle.testutils.data.Rand
import io.appmetrica.gradle.testutils.rules.MockedConstructionRule.Companion.constructionRule
import io.appmetrica.gradle.testutils.rules.MockedStaticRule.Companion.on
import io.appmetrica.gradle.testutils.rules.MockedStaticRule.Companion.staticRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.mockito.ArgumentMatcher
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.argThat
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify

internal class AppMetricaTrackerTest : CommonTest() {

    private var context: Context = mock()
    private val reporter: IReporter = mock()

    @get:Rule
    val appMetricaMockedStaticRule = staticRule<AppMetrica> {
        on { AppMetrica.getReporter(any(), any()) } doReturn reporter
    }

    private var controller: PushServiceControllerComposite = mock {
        on { transportIds } doReturn setOf(CoreConstants.Transport.FIREBASE, CoreConstants.Transport.UNKNOWN)
    }

    private var core: AppMetricaPushCore = mock {
        on { pushServiceController } doReturn controller
    }

    @get:Rule
    val appmetricaPushCoreStaticRule = staticRule<AppMetricaPushCore> {
        on { AppMetricaPushCore.getInstance(context) } doReturn core
    }

    private val preferenceManager: PreferenceManager = mock()
    private val nextEventId = 100500L

    @get:Rule
    val appMetricaTrackerEventIdGeneratorMockedConstructionRule = constructionRule<AppMetricaTrackerEventIdGenerator> {
        on { generate() } doReturn nextEventId
    }

    private val attributesMatcher = ArgumentMatcher<Map<String, Any?>> {
        assertThat(it).containsAllEntriesOf(
            mapOf(
                AppMetricaTracker.SDK_VERSION_CODE_FIELD to BuildConfig.VERSION_CODE.toString(),
                AppMetricaTracker.TRANSPORT_FIELD to
                    "[${CoreConstants.Transport.FIREBASE}, ${CoreConstants.Transport.UNKNOWN}]",
                AppMetricaTracker.EVENT_ID to nextEventId
            )
        )
        true
    }

    private val tracker: AppMetricaTracker by setUp { AppMetricaTracker(context, "Some api key", preferenceManager) }

    @Test
    fun appMetricaTrackerEventIdGenerator() {
        assertThat(appMetricaTrackerEventIdGeneratorMockedConstructionRule.constructionMock.constructed()).hasSize(1)
        assertThat(appMetricaTrackerEventIdGeneratorMockedConstructionRule.argumentInterceptor.flatArguments())
            .containsExactly(preferenceManager, "sdk")
    }

    @Test
    fun resumeSessionShouldSendResumeToReporter() {
        tracker.resumeSession()
        verify(reporter, Mockito.times(1)).resumeSession()
    }

    @Test
    fun pauseSessionShouldSendResumeToReporter() {
        tracker.pauseSession()
        verify(reporter, Mockito.times(1)).pauseSession()
    }

    @Test
    fun reportEventShouldSendEventNameToReporter() {
        val eventName = Rand.randomString()
        tracker.reportEvent(eventName)
        verify(reporter).reportEvent(eq(eventName), any<Map<String, Any?>>())
    }

    @Test
    fun reportEventShouldIncludeSdkVersionToEnvironment() {
        tracker.reportEvent(Rand.randomString())
        verify(reporter).reportEvent(any(), argThat(attributesMatcher))
    }

    @Test
    fun reportEventWithAttributesShouldIncludeSdkVersionToEnvironment() {
        tracker.reportEvent(Rand.randomString(), HashMap())
        verify(reporter).reportEvent(any(), argThat(attributesMatcher))
    }

    @Test
    fun reportEventWithAttributesShouldSendEventNameToReporter() {
        val eventName = Rand.randomString()
        tracker.reportEvent(eventName, null)
        verify(reporter).reportEvent(eq(eventName), anyOrNull<Map<String, Any?>>())
    }

    @Test
    fun reportEventWithAttributesShouldSendAttributesToReporter() {
        val attributes = mutableMapOf<String, Any?>(Rand.randomString() to Rand.randomString())
        tracker.reportEvent(Rand.randomString(), attributes)
        verify(reporter).reportEvent(any(), eq(attributes))
    }

    @Test
    fun reportErrorShouldSendMessageToReporter() {
        val message = Rand.randomString(10)
        tracker.reportError(message, null)
        verify(reporter).reportError(argThat { contains(message) }, anyOrNull<Throwable>())
    }

    @Test
    fun reportErrorShouldSendThrowableToReporter() {
        val throwable: Throwable = mock()
        tracker.reportError("null", throwable)
        verify(reporter).reportError(any(), eq(throwable))
    }

    @Test
    fun reportErrorShouldContainsSdkVersionInErrorMessage() {
        val message = Rand.randomString()
        tracker.reportError(message, null)
        verify(reporter).reportError(
            argThat<String> { contains("${AppMetricaTracker.SDK_VERSION_CODE_FIELD} = ${BuildConfig.VERSION_CODE}") },
            anyOrNull<Throwable>()
        )
    }

    @Test
    fun reportUnhandledExceptionShouldSendThrowableToReporter() {
        val throwable: Throwable = mock()
        tracker.reportUnhandledException(throwable)
        verify(reporter).reportUnhandledException(throwable)
    }
}
